### ISSUE TITLE
question to golang

### ISSUE THREAD

**Issue by geraldstanje** [VirusTotal/yara-x#144]
hi,

nice work.

i try to call yara.compile and yara.match like the github.com/VirusTotalyara-python lib...

do you have a small example which loads a yara file from disk, calls yara.compile and calls yara.match with an input string?

@plusvic does go get github.com/VirusTotal/yara-x/go@v0.4.0 also work? edit: seems yes.

how to compile the golang app using yarax for static build and what flags to set?
i get error while loading shared lib: libyara_x_capi.so.0.4 - not found for:
```
go build . -o main; ./main 
```

cc @steffenfritz

Thanks,
Gerald

**Comment by steffenfritz** [comment]
This is also something I am struggling with. I use yara-x for my file indexer FileTrove (https://github.com/steffenfritz/FileTrove) and I am fighting with Rust -> C -> Go builds and creating releases with Github Actions.

I asked in Go-Groups, they said ask Rustians. Rustians say it is something with C and Go, ask there. :)

More on topic: I think with CGO enabled you won't get a static build.

**Comment by plusvic** [comment]
@steffenfritz @geraldstanje what's your OS? Linux or Windows?

In Linux it should be relatively easy, you must install the `yara-x-capi` library first, as described in: https://virustotal.github.io/yara-x/docs/api/c/c-/

After installation, you can check if `pkg-config` is able to find the installed library:

```bash
pkg-config --libs yara_x_capi
```

The command above should produce an output like:

```
-L/usr/local/lib -lyara_x_capi
```

If that works correctly you should be able to compile your Go program without any issues. The Go compiler uses `pkg-config` for locating the required header files and libraries. The Go binary while be dynamically linked to the `yara_x_capi` library, if you want static linking you must pass `-tags static_link` to the Go compiler, like in:

```
go build -tags static_link
```

In Windows things are a bit more more complex because the `pkg-config` mechanism doesn't work, therefore the Go compiler is unable to find the header files and libraries. I need still need to figure out an easy method for building the Go library in Windows.

**Comment by geraldstanje** [comment]
hi, im using mac osx in the host and my target platform is linux amd64

**Comment by plusvic** [comment]
So, the problem is not while building the Go program, it's actually that you are unable to build the `libyara_x_capi` library, as mentioned in #145. Right?

**Comment by geraldstanje** [comment]
yeah we can close this ticket.

**Comment by steffenfritz** [comment]
As I was mentioned in this issue I will also answer, hopefully not mixing things up.

I can build the library at Linux and MacOS without problems, Windows not tested so far.

However, I have no problem to ship the library with my tool, mentioned above. At the moment I struggle with exporting the artifacts from the Github Action.

It would be pretty cool if you could create library releases when you release yara-x binaries. It could be probably pretty easy implemented (for someone more familiar with the rust-toolchain then me). I tried my luch with this action, biut without success https://github.com/VirusTotal/yara-x/blob/main/.github/workflows/golang.yaml

tl;dr: Could you also release the libraries during the release of new versions?

**Comment by plusvic** [comment]
Windows libraries will be included in the binaries produced for each release:

https://github.com/VirusTotal/yara-x/commit/cee673d8fc978bac586df2bc4c310c04d05d6181

**Comment by kmahyyg** [comment]
> [@steffenfritz](https://github.com/steffenfritz) [@geraldstanje](https://github.com/geraldstanje) what's your OS? Linux or Windows?
> 
> In Linux it should be relatively easy, you must install the `yara-x-capi` library first, as described in: https://virustotal.github.io/yara-x/docs/api/c/c-/
> 
> After installation, you can check if `pkg-config` is able to find the installed library:
> 
> pkg-config --libs yara_x_capi
> The command above should produce an output like:
> 
> ```
> -L/usr/local/lib -lyara_x_capi
> ```
> 
> If that works correctly you should be able to compile your Go program without any issues. The Go compiler uses `pkg-config` for locating the required header files and libraries. The Go binary while be dynamically linked to the `yara_x_capi` library, if you want static linking you must pass `-tags static_link` to the Go compiler, like in:
> 
> ```
> go build -tags static_link
> ```
> 
> In Windows things are a bit more more complex because the `pkg-config` mechanism doesn't work, therefore the Go compiler is unable to find the header files and libraries. I need still need to figure out an easy method for building the Go library in Windows.

If you'd like a static-linked file under Linux, no, it's not working against glibc.
I'm building under ubuntu 24.04, with the following errors:

```
root@511c3cea5a98:/opt/softsrcs/cramc/gocode# go build -trimpath -ldflags "-s -w -X \"cramc_go/common.VersionStr=$(git describe --long --dirty --tags)\" -extldflags \"-static -lm -L/lib/x86_64-linux-gnu\"" -tags static_link -o /tmp/devrev ./cmd/devreleaser
# cramc_go/cmd/devreleaser
/opt/golang/pkg/tool/linux_amd64/link: running gcc failed: exit status 1
/usr/bin/gcc -m64 -s -Wl,--build-id=0x2473fdace4bd0b10f5f7c25cd0b5aa7501dbb332 -o $WORK/b001/exe/a.out -Wl,--export-dynamic-symbol=_cgo_panic -Wl,--export-dynamic-symbol=_cgo_topofstack -Wl,--export-dynamic-symbol=crosscall2 -Wl,--export-dynamic-symbol=importCallback -Wl,--export-dynamic-symbol=matchCallback -Wl,--export-dynamic-symbol=metadataCallback -Wl,--export-dynamic-symbol=onMatchingRule -Wl,--export-dynamic-symbol=patternCallback -Wl,--export-dynamic-symbol=ruleCallback -Wl,--export-dynamic-symbol=slowestRulesCallback -Wl,--export-dynamic-symbol=tagCallback -Wl,--compress-debug-sections=zlib /tmp/go-link-320262208/go.o /tmp/go-link-320262208/000000.o /tmp/go-link-320262208/000001.o /tmp/go-link-320262208/000002.o /tmp/go-link-320262208/000003.o /tmp/go-link-320262208/000004.o /tmp/go-link-320262208/000005.o /tmp/go-link-320262208/000006.o /tmp/go-link-320262208/000007.o /tmp/go-link-320262208/000008.o /tmp/go-link-320262208/000009.o /tmp/go-link-320262208/000010.o /tmp/go-link-320262208/000011.o /tmp/go-link-320262208/000012.o /tmp/go-link-320262208/000013.o /tmp/go-link-320262208/000014.o /tmp/go-link-320262208/000015.o /tmp/go-link-320262208/000016.o /tmp/go-link-320262208/000017.o -O2 -g -L/usr/local/lib/x86_64-linux-gnu -lyara_x_capi -lgcc_s -lutil -lrt -lpthread -lm -ldl -lc -O2 -g -lpthread -no-pie -static -lm -L/lib/x86_64-linux-gnu
/usr/bin/ld: cannot find -lgcc_s: No such file or directory
collect2: error: ld returned 1 exit status
```

`libgcc_s.a` is not on my machine, but `libgcc_s.so.1` is existing. Building with/without `-L/lib/x86_64-linux-gnu` in `extldflags` won't change the error output.

**Comment by kmahyyg** [comment]
I personally believe, the only way for fully static-linked executable is link against `musl-libc`, but it comes into another question, that `libunwind` is unable to be linked statically.

And `cargo cinstall` won't generate `cdylib` against target `x86_64-unknown-linux-musl`.

**Comment by kmahyyg** [comment]
Just leave a note here in case someone would like to build a fully static-linked binary like me or even cross compile to Windows platform:

Check this GitHub action file:
https://github.com/kmahyyg/cramc/blob/v4_yarax/.github/workflows/v4_rustbuild.yml

Note: Binary on Linux is statically linked against glibc, not musl. I failed to manage to make it compile against musl.
